Driver for ITest from Bilt

edumur opened the pull request:
Allow set and get of voltage with different ramps.

FarBo commented:
Hi @edumur 
Thank you for the contribution. 
I will start the review process in a continuous basis.
I will look into why the PR is not passing the CI.

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,437 @@
+                           get_parser=float
+                           )
+        
+        self.add_parameter('v_and_trig',
```
a curiosity/design question - why is it such that the `v` parameter that has the more user-friendly name (similary to other qcodes drivers with `volt` and `voltage` parameters) does NOT actually set the voltage to that value? i'd expect that the v parameter DOES set the voltage and there's some target_v parameter that allows to tell the instrument about one's intent to set the voltage to some value.

another remark - once a voltage is being set via v_and_trig parameter, shouldn't parameter v be updated to have an invalid cache (see `.cache.invalidate` for more info), or even set to some neautral value?

edumur replied:
Hi,

Thanks for the remarks.
The trigger story is relative to how the ITest is built.
1. you set a ramp mode.
2. you set a ramp speed.
3. you set a target voltage.
4. you trigger the ramping.

However I discussed with a colleague and he had input that I am going to integrate and come back with a new version.
I will also change the naming to target_voltage which is maybe more meaningful.

edumur commented:
Hi all,

I remove the trigger thing so now there is only a simple voltage parameter.
I added a fake parameter called "synchronous" which block or not the communication with the instrument.
I updated the example accordingly.

edumur commented:
Could someone review this driver?
I would like to propose another one after this one?

Thanks!

FarBo commented:
Hi edmur

Thanks for the reminder. I will review it this week.

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,467 @@
+        self._synchronous[chan_id] = synchronous
+
+
+    def _get_chan_synchronous(self, chan: int) -> None:
```
You are returning self._synchronous[chan_id], but saying return is None.  Please fix ths

edumur replied:
Corrected.

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,467 @@
+        self.write(chan_id + 'chan:name "{}"'.format(name))
+
+
+    def chan_to_ic(self, chan:int) -> Tuple[int]:
```
What are the types of 'i' and 'c'?  I think 'i' is a float and 'c' is an int. If yes, then the return type should be Tuple[float, int]

edumur replied:
Corrected.

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,467 @@
+        if state == '1':
+            return True
+        elif state == '0':
+            return False
```
What do you expect if state is not '1' or '0'? can they be other than these two options?

edumur replied:
I do not expect `state` to be anything different.
However I added an `else` statement with a error in case of.

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,467 @@
+        self.write('{}VOLT:SLOP {:.8f}'.format(chan_id, slope))
+
+
+    def _get_ramp_slope(self, chan:int) -> float:
```
The return shoudln't be str?

edumur replied:
Yes indeed !

FarBo commented:
I put some comments to solve the CI issues. Let's first correct the PR to pass the CI, then I will review the content.

edumur commented:
Done, thanks for your input!

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,469 @@
+                           docstring='Voltage of the channel in volt.',
+                           get_cmd=partial(self._parent._get_voltage, chan_num),
+                           set_cmd=partial(self._parent._set_voltage, chan_num),
+                           get_parser=float
```
Doesn't it need any validator for set values?
It can prevent setting out of limit values in software level.

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,469 @@
+                           get_cmd=partial(self._parent._get_ramp_slope, chan_num),
+                           get_parser=float,
+                           set_cmd=partial(self._parent._set_ramp_slope, chan_num),
+                           set_parser=float
```
validators here?

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,469 @@
+                           set_cmd=partial(self._parent._set_chan_range, chan_num),
+                           set_parser=float,
+                           get_cmd=partial(self._parent._get_chan_range, chan_num),
+                           get_parser=float)
```
validators here?

FarBo replied:
Are you sure no validators needed here?

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,469 @@
+                           get_parser=str
+                           )
+        
+        self.add_parameter('range',
```
maybe changing the name to v_range or vrange makes more sense?

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,469 @@
+                      synchronous:bool=True,
+                      synchronous_delay:float=1,
+                      synchronous_threshold:float=1e-5,
+                      **kwargs) -> None:
```
**kwargs: Any

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,469 @@
+# Loick Le Guevel, 2019
+# Etienne Dumur <etienne.dumur@gmail.com>, 2021
+
+from typing import Union, Tuple
```
please add Any here

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,469 @@
+        """
+        super().__init__(name, address=address,
+                               terminator='\n',
+                               device_clear=False)
```
You can add **kwargs to this super.__init__ as well?

FarBo commented:
Thank you for the fix.
I also made some comments in the code. Also, I have a question:
Can pos and neg states be set independently? I ask this because I don't know about the instrument, but I am thinking that setting positive state should disable negative state and vice versa. What do you think?

edumur commented:
I added your correction.

Concerning your question, I tested both pos and neg request on the device directly and setting one does not disable the other.

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+                           get_cmd=partial(self._parent._get_voltage, chan_num),
+                           get_parser=float,
+                           set_cmd=partial(self._parent._set_voltage, chan_num),
+                           set_parser=vals.Numbers(-50, 50)
```
You can simply say: vals=vals.Numbers(-50, 50)

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+                           get_cmd=partial(self._parent._get_ramp_slope, chan_num),
+                           get_parser=float,
+                           set_cmd=partial(self._parent._set_ramp_slope, chan_num),
+                           set_parser=vals.Numbers(0, 1)
```
same here

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+                           get_cmd=partial(self._parent._get_chan_synchronous, chan_num),
+                           get_parser=bool,
+                           set_cmd=partial(self._parent._set_chan_synchronous, chan_num),
+                           set_parser=bool)
```
Are you sure no validators needed here?

FarBo commented on docs/examples/ITest.ipynb:
```diff
@@ -0,0 +1,199 @@
+    }
+   ],
+   "source": [
+    "dac.ch04.v(0.5), dac.ch04.v()"
```
Perhaps it's better to split each command in one cell, because in this way, it returns None for dac.ch04.v(0.5), which is misleading.

edumur commented:
Done!

FarBo commented:
@edumur 
Thank you for the commits. Could you mark the comments as resolved?

@astafan8 
Could you have a final look?

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+                           get_cmd=partial(self._parent._get_output_function, chan_num),
+                           get_parser=str,
+                           set_cmd=partial(self._parent._set_output_function, chan_num),
+                           set_parser=str
```
is it true that the output_mode can only have a certain strings as values? for example, only "ramp" and "non-ramp". If so, then it will be very useful for users and for correctness to use an Enum validator with those strings:
```suggestion
                           set_parser=str,
                           vals=vals.Enum("ramp", "non-ramp")
```

FarBo replied:
This is recommended.

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+                           get_cmd=partial(self._parent._get_chan_range, chan_num),
+                           get_parser=float)
+        
+        self.add_parameter('state',
+                           docstring='State of the channel {on, off}.',
+                           get_cmd=partial(self._parent._get_chan_state, chan_num),
+                           get_parser=bool,
+                           set_cmd=partial(self._parent._set_chan_state, chan_num),
+                           vals=vals.OnOff())
```
as per this section in the docs https://qcodes.github.io/Qcodes/examples/writing_drivers/Creating-Instrument-Drivers.html#On/Off-parameters , i highly recommend to implement this parameter as suggested below and adjust `_get_chan_state`/`_set_chan_state` methods to accept "on"/"off" as input (and not perform the conversion to them from a boolean):
```suggestion
        self.add_parameter('enabled',
                           docstring='Whether the channel is enabled or not',
                           get_cmd=partial(self._parent._get_chan_state, chan_num),
                           set_cmd=partial(self._parent._set_chan_state, chan_num),
                           val_mapping=create_on_off_val_mapping(on_val="on", off_val="off")
```

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+        self.add_function('stop',
+                          call_cmd=lambda: self._parent._set_chan_state(self.chan_num, 0))
+    
+        self.bilt_name(f'Chan{chan_num:02d}')
```
this can be included into the definition of the bilt_name parameter with `initial_value=f'Chan{chan_num:02d}'` argument

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+        self.add_function('start',
+                          call_cmd=lambda: self._parent._set_chan_state(self.chan_num, 1))
+        
+        self.add_function('stop',
```
please refrain from using functions unless it's really needed, see here https://qcodes.github.io/Qcodes/examples/writing_drivers/Creating-Instrument-Drivers.html#Functions, likely simple bound methods would do the job

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+                           docstring='Mode of the output {exp, ramp}.',
+                           get_cmd=partial(self._parent._get_output_function, chan_num),
+                           get_parser=str,
+                           set_cmd=partial(self._parent._set_output_function, chan_num),
```
so far as i see the implementation of the _set_output_function method, apparently output_mode can be also supplies as a float or an int? if that is so, then this parameter should allow that, right? or should that feature be removed from _set_output_function ? if the parameter needs to be able to not only accept a few selected strings but selected numbers, then those numbers can also be added to the `vals.Enum`. or if the parameter is supposed to accept a range of numbers then the `vals.MultiType` validator can be used: `vals.MultiType(vals.Enum("ramp", "foo"), vals.Numbers(0, 2.5))`

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+        super().__init__(channels, param_name, *args, **kwargs)
+
+    def trigger(self,*args,**kwargs):
+        print(args,kwargs)
```
is this really needed? :) having a trigger method that just prints its arguments?)) if it's not needed, it's better to remove in order not to confuse the future readers of this code

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+        self.idn = self.get_idn()
+        self.num_chans = num_chans
+        self.chan_range = range(1,self.num_chans+1)
+        # Fake parameter allowing to control the synchronicity of every channels
```
is there a reason for these being init-only private attributes as opposed to proper parameters that users can change when using the instrument? just in case, a reminder, parameters can have set_cmd=None and get_cmd=None which makes them "in-memory" parameters, so they don't talk to the instrument, but still maintain a value. and then instead of maintaining the `self._synchronous` dictionary on the instrument level, this information can be conveniently maintained on the channel level

edumur replied:
Hum, could point.
I could change it and make in an 'in-memory' thing at the channel level but then how would I access it at the instrument level, for example here:
```python
    def _set_voltage(self, chan:int,
                           v_set:float) -> None:
        """
        Set cmd for the chXX_v parameter
        
        Args:
            chan: The 1-indexed channel number
            v_set: The target voltage
        """
        chan_id = self.chan_to_id(chan)
        self.write('{}VOLT {:.8f}'.format(chan_id, v_set))
        self.write(chan_id + 'TRIG:INPUT:INIT')
        if self._get_chan_synchronous(chan):
            v = self._get_voltage(chan)
            while abs(v_set - v)>=self._synchronous_threshold:
                sleep(self._synchronous_delay)
                v = self._get_voltage(chan)
```

Thanks.

astafan8 replied:
since the instrument is aware of all the channels and they can be accessed by their name(?), then `self.<chanX>.synchronous_delay()` will just work.

now that i look at this method, it seems to me that it actually also should go inside the channel because there is nothing in this method that depends on the instrument itself. the write/ask method calls on the channel level forward to those on the instrument level, so that you don't have to worry about

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+        elif mode == 1:
+            return 'ramp'
+        else:
+            return 'ERROR'
```
shall this raise a proper exception with useful error message? :) otherwise it's less likely that users will understand what is wrong :) something like, `raise ValueError(f"Got unexpected output function mode {mode}")`

astafan8 commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -0,0 +1,475 @@
+    def _set_chan_neg_sat(self, chan:int,
+                                neg_sat: Union[int, str]) -> None:
+        chan_id = self.chan_to_id(chan)
+        if isinstance(neg_sat,(int,float)):
```
whenever a function accepts a union of numbers and strings, the types should be changed to `Union[float, str]` becasue floats are allowed as well (right?)

astafan8 reviewed (approved):
I left some comments, but only a few of them are critical (for example, having the correct types in methods signatures); it's up to you @edumur and @FarBo to judge what's worth changing and what can wait for the future

FarBo commented:
Thanks Mikhail, I think most of your comments are recommended and it's up to @edumur how he wants to implement them.

@edumur, if you like to consider these comments, please implement them and I will have a final review and approve it. Then we can merge it. Let me know what you prefer.

edumur commented:
@FarBo I will work on the nice comments of @astafan8 ;) .

edumur commented:
I, normally, integrated all the comments 🤞.

FarBo commented:
@edumur 
Could you resolve the comments that are not open?

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -101,22 +107,47 @@ def __init__(self, parent: Instrument,
 
-        self.add_parameter('synchronous',
-                           get_cmd=partial(self._parent._get_chan_synchronous, chan_num),
+        self.add_parameter('synchronous_enable',
```
If you want the instrument is initilized with this parameter, rather than explicitly putting it in the __init__ as a class argument and the creating only readable parameter later, you can create a settable parameter and get it under __init__ like:
def __init__(...):
  .
  .
  .
  self.synchronous_enable()
And don't forget to change synchronous_enable to self.synchronous_enable() anywhere else you used it.

Then if you want it to be initilized with True, you just simply set it to its initial value:
 
```
self.add_parameter('synchronous_enable',
                           docstring='Is the channel in synchronous mode.',
                           get_cmd=None,
                           get_parser=bool,
                           set_cmd=partial(self._parent._set_chan_synchronous, chan_num),
                           initial_value=True
                           vals=vals.Bool())
```

what you have done seems duplication of creating an argument and excatly the same parameter, which is set in init and get as a parameter.

edumur replied:
I wanted the user to have the possibility to change this parameter at the initialization of the device.
Is there another way of doing it?

FarBo replied:
If you want to be forced at initialization, then there should not be any default value in it
Then you have it as an argument and pass it to the parameter under init, for example:

Def init(sync: bool):
  Self.enabled_synchronous(sync)


And you remove the initial value from the parameter

edumur replied:
I guess it is done.
Please check if it is done in the way you suggested.

FarBo commented:
@edumur 

Thanks for the update. I only have one comment, If it is clarified, we can close this PR.

FarBo commented on qcodes_contrib_drivers/drivers/Bilt/ITest.py:
```diff
@@ -21,18 +21,11 @@ class iTestChannel(InstrumentChannel):
         Args:
             parent: The instrument to which the channel is attached.
             name: The name of the channel.
-            chan_num: The number of the channel in question
```
Perhaps chan_num explanation should stay :)

edumur replied:
oops
